File: Season 01 - CS50/Week 06 - Python/Credit/credit.py
# Prompt user for credit card
def get_int(prompt):
    while True:
        try:
            return int(input(prompt))
        except ValueError:
            pass


# Calculate checksum
def calculate(number):
    checksum = 0
    length = 0
    while number > 0:
        checksum += number % 10
        number = int(number / 10)
        length += 1
        if number > 0:
            tmp = 2 * (number % 10)
            number = int(number / 10)
            length += 1
            while tmp > 0:
                checksum += tmp % 10
                tmp = int(tmp / 10)
    return checksum, length


# Validate the card and print
# AMEX, VISA or MASTERCARD
def validate(number, checksum, length):
    d = {
        34: {"length": [15], "name": "AMEX"},
        37: {"length": [15], "name": "AMEX"},
        51: {"length": [16], "name": "MASTERCARD"},
        52: {"length": [16], "name": "MASTERCARD"},
        53: {"length": [16], "name": "MASTERCARD"},
        54: {"length": [16], "name": "MASTERCARD"},
        55: {"length": [16], "name": "MASTERCARD"},
        4: {"length": [13, 16], "name": "VISA"},
    }
    # Trim the credit card to get the prefix
    prefix_digits = 2
    prefix = number
    for i in range(length - prefix_digits):
        prefix = int(prefix / 10)
    if checksum % 10 != 0:
        print("INVALID")
    else:
        # Multiple prefix are tried
        for i in range(1, prefix_digits + 1, 1):
            try:
                if length in d[prefix]["length"]:
                    print(d[prefix]["name"])
                    return
            except KeyError:
                pass
            # Remove the last digit of the prefix
            prefix = int(prefix / 10)
        # None of the prefix tried was found in the dictionary
        print("INVALID")


def main():
    creditCard = get_int("Number: ")
    checksum, length = calculate(creditCard)
    validate(creditCard, checksum, length)


main()

File: Part 01 - CS50/Week 01 - C/Credit/credit.c
#include "cs50.h"
#include <stdio.h>

int cardLength;
int cardParity;
int cardChecksum;
int checksumAdd;
long moduloFactor = 1;
int main(void)
{
    // Prompt for credit card input
    const long cardNumber = get_long("Enter your credit card number: ");
    // Calculate checksum
    while (cardNumber > moduloFactor)
    {
        // Modulo 10
        cardLength++;
        cardParity = cardLength % 2;
        checksumAdd = (cardNumber / moduloFactor) % 10;
        if (cardParity == 1)
        {
            cardChecksum = checksumAdd + cardChecksum;
        }
        else if (cardParity == 0 && checksumAdd < 5)
        {
            cardChecksum = 2 * checksumAdd + cardChecksum;
        }
        else if (cardParity == 0 && checksumAdd >= 5)
        {
            cardChecksum = 2 * checksumAdd % 10 + 2 * checksumAdd / 10 % 10 + cardChecksum;
        }
        moduloFactor *= 10;
        printf("%i\n", cardChecksum);
    }

    printf("%i\n", cardLength);
    int cardPrefix1 = (cardNumber * 10) / moduloFactor;
    int cardPrefix2 = (cardNumber * 100) / moduloFactor;
    if (cardChecksum % 10 != 0)
    {
        printf("INVALID\n");
    }
    else if ((cardPrefix2 == 34 && cardLength == 15) || (cardPrefix2 == 37 && cardLength == 15))
    {
        printf("AMEX\n");
    }
    else if ((cardPrefix2 == 51 && cardLength == 16) || (cardPrefix2 == 52 && cardLength == 16) ||
             (cardPrefix2 == 53 && cardLength == 16) || (cardPrefix2 == 54 && cardLength == 16) ||
             (cardPrefix2 == 55 && cardLength == 16))
    {
        printf("MASTERCARD\n");
    }
    else if ((cardPrefix1 == 4 && cardLength == 13) || (cardPrefix1 == 4 && cardLength == 16))
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }
}
